Game/marvin.cpp

```diff
@@ -6,132 +6,130 @@
+  commandHandlers["toggle camdebug"] = CommandFunction(&Marvin::handleToggleCamDebug);
+  commandHandlers["toggle camera"] = CommandFunction(&Marvin::handleToggleCamera);
+  commandHandlers["insert"] = CommandFunction(&Marvin::handleInsert);
+  }
 
-Marvin::Marvin() {
-  cmd = std::vector<Cmd>{
-    {"cheat full",        C_CheatFull},
+void Marvin::autoComplete(std::string& inputString) {
+  inputString = StringUtil::trimFront(inputString, ' ');
+  CommandMap::iterator it;
 
-    {"camera autoswitch", C_CamAutoswitch},
-    {"camera mode",       C_CamMode},
-    {"toogle camdebug",   C_ToogleCamDebug},
-    {"toogle camera",     C_ToogleCamera},
-    };
-  }
+  size_t matchCount = 0;
+  std::string matchCommand;
 
-Marvin::CmdVal Marvin::recognize(const std::string& v) {
-  char clean[256] = {};
-
-  for(size_t i=0, r=0; r<255 && i<v.size(); ++i) {
-    char c = v[i];
-    if('A'<=c && c<='Z')
-      c = char(c+'a'-'A');
-    if(r==0 && c==' ')
-      continue;
-    if(c==' ') {
-      while(v[i+1]==' ')
-        ++i;
-      }
+  for(it = commandHandlers.begin(); it != commandHandlers.end(); it++) {
+    const std::string command(it->first);
 
-    clean[r] = c;
-    ++r;
-    }
+    if(inputString.length() < command.length()) {
+      std::string restOfCommand( command.substr(inputString.length()));
+      std::string::size_type spacePos = restOfCommand.find(" ");
+      std::string commandToken(command);
 
-  size_t cmdLen = std::strlen(clean);
-  size_t prefix = cmdLen;
-  if(prefix==0)
-    return C_None;
+      if( spacePos != std::string::npos ) {
+        commandToken = command.substr(0, inputString.length() + spacePos);
+      }
 
-  while(prefix>0) {
-    if(clean[prefix]==' ')
-      break;
-    --prefix;
-    }
 
-  size_t  cnt           = 0;
-  size_t  suggestionLen = 0;
-  Cmd     suggestion;
-
-  for(auto& i:cmd) {
-    const size_t len = std::strlen(i.cmd);
-    if(len<prefix)
-      continue;
-    if(std::memcmp(clean,i.cmd,prefix)!=0)
-      continue;
-
-    if(cmdLen<=len && std::memcmp(clean,i.cmd,cmdLen)==0) {
-      size_t len = cmdLen;
-      while(i.cmd[len]!='\0' && i.cmd[len]!=' ')
-        ++len;
-      if(len!=suggestionLen || std::memcmp(suggestion.cmd,i.cmd,len)!=0) {
-        suggestion    = i;
-        suggestionLen = len;
-        cnt++;
+      if(commandToken.substr(0,inputString.length()) == inputString) {
+        if(matchCommand != commandToken) {
+          matchCommand = commandToken;
+          matchCount++;
         }
       }
     }
+  }
 
-  if(cnt==0)
-    return C_Invalid;
+  if(matchCount == 1) {
+    inputString = matchCommand;
+  }
+  }
 
-  if(cnt!=1)
-    return C_Incomplete;
+bool Marvin::exec(const std::string& inputString) {
+  bool ret = false;
+  size_t matchCount = 0;
+  std::string matchCommand("");
 
-  if(suggestion.cmd[cmdLen]=='\0')
-    return CmdVal{suggestion,cmdLen};
+  std::string trimedInputString = StringUtil::trimFront(inputString, ' ');
```

Comment:
The command recognition seems to be duplicated across 2 functions: `Marvin::exec` and `Marvin::autoComplete`

Reply:
Possibly, will need to look into this as part of the discussion about the recognize function.